core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -326,4 +322,8 @@ public Version indexVersionCreated() {
     public QueryParseContext parseContext() {
         return this.parseContext;
     }
+
+    public Client client() {
```

Comment:
you don't need to expose the client to everybody. You can contain which objects depend on the client simply having a handleTermsLookup method or something along those lines in IndexQueryParseService. That one will use the client but that way the client will not be exposed through public getter.